File: backend/app/services/gmo_payment.py
"""
GMOペイメントゲートウェイ リンクタイプPlus クライアント

ハッシュ型接続方式:
1. 決済パラメータJSON → Base64エンコード（α）
2. α + ShopPass → SHA256ハッシュ（γ）
3. α.γ → 決済URL生成
"""

import base64
import hashlib
import json
import logging
import uuid

from app.config import settings

logger = logging.getLogger(__name__)


def generate_order_id() -> str:
    """一意な注文IDを生成する（GMO PGの制約: 英数字とハイフンのみ）"""
    return f"ENC-{uuid.uuid4().hex[:16].upper()}"


def build_payment_url(
    order_id: str,
    amount: int,
    overview: str = "アンコール決済",
) -> str:
    """
    GMOリンクタイプPlusのハッシュ型決済URLを生成する。

    Args:
        order_id: 注文ID（一意、英数字・ハイフンのみ）
        amount: 決済金額（円）
        overview: 取引概要

    Returns:
        GMO決済ページのURL
    """
    params = {
        "configid": settings.gmo_config_id,
        "transaction": {
            "OrderID": order_id,
            "Amount": str(amount),
            "Overview": overview,
        },
    }

    json_str = json.dumps(params, ensure_ascii=False, separators=(",", ":"))

    # Step 1: Base64エンコード → α
    alpha = base64.b64encode(json_str.encode("utf-8")).decode("ascii")

    # Step 2: α + ShopPass → β, SHA256(β) → γ
    beta = alpha + settings.gmo_shop_pass
    gamma = hashlib.sha256(beta.encode("utf-8")).hexdigest()

    # Step 3: α.γ → ε
    epsilon = f"{alpha}.{gamma}"

    # 決済URL
    url = f"{settings.gmo_link_url}/v1/plus/{settings.gmo_shop_id}/checkout/{epsilon}"

    logger.info(f"GMO決済URL生成: order_id={order_id}, amount={amount}")
    return url


def verify_result_notification(
    order_id: str,
    amount: str,
    shop_id: str,
    hash_value: str,
) -> bool:
    """
    GMOからの結果通知のハッシュ値を検証する。

    GMOは結果通知時に ShopID + OrderID + Amount + HashKey の
    SHA256ハッシュを送信する。これを検証して改ざんを検出する。
    """
    if not settings.gmo_result_hash_key:
        logger.warning("GMO結果通知ハッシュキーが未設定です")
        return True  # 開発環境では検証をスキップ

    raw = f"{shop_id}{order_id}{amount}{settings.gmo_result_hash_key}"
    expected = hashlib.sha256(raw.encode("utf-8")).hexdigest()

    if expected != hash_value:
        logger.error(
            f"GMO結果通知ハッシュ不一致: order_id={order_id}, "
            f"expected={expected}, received={hash_value}"
        )
        return False

    return True

File: frontend/src/lib/gmo-payment.ts
/**
 * GMOペイメントゲートウェイ リンクタイプPlus（ハッシュ型）
 *
 * フロー:
 * 1. 決済パラメータJSON → Base64エンコード（α）
 * 2. α + ShopPass → SHA256ハッシュ（γ）
 * 3. α.γ → 決済URL生成
 */

function getEnv(key: string): string {
  const value = import.meta.env[key] || '';
  return value;
}

/** 一意な注文IDを生成する（GMO制約: 英数字とハイフンのみ） */
export function generateOrderId(): string {
  const hex = Array.from(crypto.getRandomValues(new Uint8Array(8)))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')
    .toUpperCase();
  return `ENC-${hex}`;
}

/** SHA256ハッシュを計算する */
async function sha256(message: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(message);
  const hashBuffer = await crypto.subtle.digest('SHA-256', data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
}

/** Base64エンコード */
function toBase64(str: string): string {
  const encoder = new TextEncoder();
  const bytes = encoder.encode(str);
  let binary = '';
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

/**
 * GMOリンクタイプPlusのハッシュ型決済URLを生成する
 */
export async function buildPaymentUrl(
  orderId: string,
  amount: number,
  overview: string = 'アンコール決済',
): Promise<string> {
  const shopId = getEnv('GMO_SHOP_ID');
  const shopPass = getEnv('GMO_SHOP_PASS');
  const configId = getEnv('GMO_CONFIG_ID');
  const linkUrl = getEnv('GMO_LINK_URL') || 'https://stg.link.mul-pay.jp';

  const params = {
    configid: configId,
    transaction: {
      OrderID: orderId,
      Amount: String(amount),
      Overview: overview,
    },
  };

  const jsonStr = JSON.stringify(params);

  // Step 1: Base64エンコード → α
  const alpha = toBase64(jsonStr);

  // Step 2: α + ShopPass → β, SHA256(β) → γ
  const beta = alpha + shopPass;
  const gamma = await sha256(beta);

  // Step 3: α.γ → ε
  const epsilon = `${alpha}.${gamma}`;

  return `${linkUrl}/v1/plus/${shopId}/checkout/${epsilon}`;
}

/**
 * GMOからの結果通知のハッシュ値を検証する
 */
export async function verifyResultNotification(
  orderId: string,
  amount: string,
  shopId: string,
  hashValue: string,
): Promise<boolean> {
  const hashKey = getEnv('GMO_RESULT_HASH_KEY');

  if (!hashKey) {
    // 開発環境では検証をスキップ
    console.warn('GMO_RESULT_HASH_KEY is not set, skipping hash verification');
    return true;
  }

  const raw = `${shopId}${orderId}${amount}${hashKey}`;
  const expected = await sha256(raw);

  if (expected !== hashValue) {
    console.error(`GMO hash mismatch: order=${orderId}, expected=${expected}, received=${hashValue}`);
    return false;
  }

  return true;
}
